### PR TITLE
homework is now ready yet

### PR DESCRIPTION
**Pull request by dezuz**

### CODE REVIEW

File: src/main/java/com/mateacademy/configuration/DataConfig.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    private static final String PROP_HIBERNATE_DIALECT = "spring.jpa.properties.hibernate.dialect";
+    private static final String PROP_HIBERNATE_SHOW_SQL = "spring.jpa.show-sql";
+    private static final String PROP_HIBERNATE_HBM2DDL_AUTO = "spring.jpa.hibernate.ddl-auto";
+
+    @Resource
+    private Environment env;
+
+    @Bean
+    public DataSource dataSource() {
+        DriverManagerDataSource dataSource = new DriverManagerDataSource();
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/configuration/DataConfig.java]: What do you think about using overload constructor
DriverManagerDataSource(String url, String username, String password)?
https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/jdbc/datasource/DriverManagerDataSource.html

File: src/main/java/com/mateacademy/configuration/DataConfig.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        entityManagerFactoryBean.setJpaVendorAdapter(new HibernateJpaVendorAdapter());
+        entityManagerFactoryBean.setPackagesToScan("com.mateacademy.model");
+
+        Properties jpaProperties = new Properties();
+        jpaProperties.setProperty(PROP_HIBERNATE_DIALECT, env.getRequiredProperty(PROP_HIBERNATE_DIALECT));
+        jpaProperties.put(PROP_HIBERNATE_SHOW_SQL, env.getRequiredProperty(PROP_HIBERNATE_SHOW_SQL));
+        jpaProperties.setProperty(PROP_HIBERNATE_HBM2DDL_AUTO, env.getRequiredProperty(PROP_HIBERNATE_DIALECT));
+        entityManagerFactoryBean.setJpaProperties(jpaProperties);
+
+        entityManagerFactoryBean.setJpaProperties(jpaProperties);
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/configuration/DataConfig.java:63]: Duplicate (see line 63)

File: src/main/java/com/mateacademy/configuration/DataConfig.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        dataSource.setUrl(env.getRequiredProperty(PROP_DATABASE_URL));
+        dataSource.setUsername(env.getRequiredProperty(PROP_DATABASE_USERNAME));
+        dataSource.setPassword(env.getRequiredProperty(PROP_DATABASE_PASSWORD));
+
+        return dataSource;
+    }
+
+    @Bean
+    public LocalContainerEntityManagerFactoryBean entityManagerFactory() {
+        LocalContainerEntityManagerFactoryBean entityManagerFactoryBean = new LocalContainerEntityManagerFactoryBean();
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/configuration/DataConfig.java]: What do you think about using builder here?
https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.Builder.html

File: src/main/java/com/mateacademy/configuration/DataConfig.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        entityManagerFactoryBean.setDataSource(dataSource());
+        entityManagerFactoryBean.setPersistenceProviderClass(HibernatePersistenceProvider.class);
+        entityManagerFactoryBean.setJpaVendorAdapter(new HibernateJpaVendorAdapter());
+        entityManagerFactoryBean.setPackagesToScan("com.mateacademy.model");
+
+        Properties jpaProperties = new Properties();
+        jpaProperties.setProperty(PROP_HIBERNATE_DIALECT, env.getRequiredProperty(PROP_HIBERNATE_DIALECT));
+        jpaProperties.put(PROP_HIBERNATE_SHOW_SQL, env.getRequiredProperty(PROP_HIBERNATE_SHOW_SQL));
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/configuration/DataConfig.java]: Please use  `setProperty`.
https://stackoverflow.com/questions/30381563/putting-objects-into-java-util-properties

File: src/main/java/com/mateacademy/controller/LoginController.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+
+import java.security.Principal;
+
+@Controller
+public class LoginController {
+    private UserService userService;
+
+    public LoginController(UserService userService) {
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/controller/LoginController.java]: How did you inject userService into the controller?
@RequiredArgsConstructor(onConstructor = @__(@Autowired)) could help you 
https://www.baeldung.com/spring-injection-lombok

File: src/main/java/com/mateacademy/controller/LoginController.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        this.userService = userService;
+    }
+
+    @GetMapping(value = "/login")
+    public String login(Principal principal) {
+        return "login.html";
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/controller/LoginController.java]: Should we add an extension for the view?

File: src/main/java/com/mateacademy/util/SecurityUtil.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.security.core.userdetails.User;
+import org.springframework.stereotype.Component;
+
+import java.security.Principal;
+import java.util.Objects;
+import java.util.Optional;
+
+@Component
+@Slf4j
+public class SecurityUtil {
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/util/SecurityUtil.java]: https://www.vojtechruzicka.com/avoid-utility-classes/
It is not utility class. Please rename it.
SecurityUtil ->SecurityService for example

**Review by Anderwerty — COMMENTED** [review]
1. Please finish your project
2. Resolve comments

File: src/main/java/com/mateacademy/controller/LoginController.java (hunk 5 above)
Comment (dezuz) [reply src/main/java/com/mateacademy/controller/LoginController.java]: using constructor, line 13

Comment (Anderwerty) [reply src/main/java/com/mateacademy/controller/LoginController.java]: You don't use annotation @Autowired or @Inject. So how can it be injected?

File: src/main/java/com/mateacademy/model/UserModel.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
+package com.mateacademy.model;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.experimental.Accessors;
+
+import javax.persistence.*;
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/model/UserModel.java]: Please don't use a wild card in imports

File: src/main/java/com/mateacademy/configuration/SecurityContext.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.*;
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/configuration/SecurityContext.java]: Please don't use a wild card in import.

File: src/main/java/com/mateacademy/configuration/SecurityContext.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+                .logout()
+                .logoutUrl("/logout")
+                .permitAll()
+                .and()
+                .exceptionHandling().accessDeniedHandler(accessDeniedHandler);
+    }
+
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        String idForEncode = "bcrypt";
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/configuration/SecurityContext.java:57]: Please use constant here.

File: src/main/java/com/mateacademy/configuration/SecurityContext.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        Map encoders = new HashMap<>();
+        encoders.put(idForEncode, new BCryptPasswordEncoder());
+        encoders.put("noop", NoOpPasswordEncoder.getInstance());
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/configuration/SecurityContext.java]: The same. For strings  "noop", "pbkdf2","scrypt",... 
Please move them to constants.

File: src/main/java/com/mateacademy/controller/UserController.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.servlet.ModelAndView;
+
+import java.security.Principal;
+
+@Controller
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class UserController {
+    private UserService userService;
+    private SecurityService utils;
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/controller/UserController.java]: utils -> securityService 
(Please don't copy your teammate's code)

File: src/main/java/com/mateacademy/model/UserModel.java (hunk 8 above)
Comment (Anderwerty) [reply src/main/java/com/mateacademy/model/UserModel.java]: You have not resolved this comment yet.

File: src/main/java/com/mateacademy/service/SecurityService.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import java.security.Principal;
+import java.util.Objects;
+import java.util.Optional;
+
+@Component
+@Slf4j
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class SecurityService {
+
+    private UserService userService;
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/service/SecurityService.java]: Please add modifier final. The same for the other field.

File: src/main/java/com/mateacademy/service/SecurityService.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    private PasswordEncoder passwordEncoder;
+
+    public UserModel getUser(Principal principal) {
+        UsernamePasswordAuthenticationToken token = (UsernamePasswordAuthenticationToken) principal;
+        User user = (User)token.getPrincipal();
+
+        String password = "11111";
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/service/SecurityService.java]: Why did you hard code?

File: src/main/java/com/mateacademy/service/UserServiceImp.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import java.util.Optional;
+
+@Service
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class UserServiceImp implements UserService {
+
+    private UserRepository userRepository;
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/service/UserServiceImp.java]: Add modifier final.

File: src/main/java/com/mateacademy/service/UserServiceImp.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    private PasswordEncoder passwordEncoder;
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/service/UserServiceImp.java]: The same

File: src/main/java/com/mateacademy/service/UserServiceImp.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import java.util.Optional;
+
+@Service
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class UserServiceImp implements UserService {
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/service/UserServiceImp.java:13]: Please add tests for this class. 
https://stackoverflow.com/questions/2288386/mockito-how-to-test-my-service-with-mocking

File: src/main/java/com/mateacademy/service/UserServiceImp.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+    private PasswordEncoder passwordEncoder;
+
+    @Override
+    public Optional<UserModel> getById(Long id) {
+        return userRepository.findById(id);
+    }
+
+    @Override
+    public Optional<UserModel> update(UserModel user) {
+        return Optional.of(userRepository.save(user));
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/service/UserServiceImp.java:24]: What if repository return null?

File: src/main/java/com/mateacademy/service/UserServiceImp.java [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Override
+    public Optional<UserModel> create(UserModel user) {
+        return Optional.of(userRepository.save(user));
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/service/UserServiceImp.java:29]: What if repository return null?

File: src/main/resources/application.properties [hunk 20]
```diff
@@ -1,8 +1,9 @@
...
 spring.datasource.password=root
-spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
 
 spring.jpa.hibernate.ddl-auto=create
 spring.jpa.show-sql=true
-spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.MySQL5Dialect
\ No newline at end of file
+spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.MySQL5Dialect
+
+spring.thymeleaf.cache=false
```
Comment (Anderwerty) [inline src/main/resources/application.properties]: Add a blank line at the end of the file.

File: src/main/resources/templates/admin.html [hunk 21]
```diff
@@ -0,0 +1,9 @@
+<html>
+<head>
+    <title>Welcome</title>
+</head>
+<body>
+
+<h1> Welcome !</h1>
+</body>
+</html>
```
Comment (Anderwerty) [inline src/main/resources/templates/admin.html]: Please add a blank line at the end of the file.

**Review by Anderwerty — CHANGES_REQUESTED** [review]

File: src/main/java/com/mateacademy/model/UserModel.java (hunk 8 above)
Comment (dezuz) [reply src/main/java/com/mateacademy/model/UserModel.java]: I resolved this after commit, so I decided to push this with next commit

File: src/main/java/com/mateacademy/controller/UserController.java (hunk 12 above)
Comment (dezuz) [reply src/main/java/com/mateacademy/controller/UserController.java]: Я і не пробував, Саша робив приклад програми на лекції, я звідти і брав частини коду.

File: src/test/java/com/mateacademy/service/UserServiceImpTest.java [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+import com.mateacademy.model.Role;
+import com.mateacademy.model.UserModel;
+import com.mateacademy.repository.UserRepository;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+public class UserServiceImpTest {
```
Comment (Anderwerty) [inline src/test/java/com/mateacademy/service/UserServiceImpTest.java:12]: Please read https://javacodehouse.com/blog/mockito-tutorial and fix it

**Review by Anderwerty — APPROVED** [review]

**Pull request merged**